### DISCUSSION TITLE
Can glob directory exclusion work at the same time as extension exclusion?

### DISCUSSION THREAD

**Discussion by tuxayo** [BurntSushi/ripgrep#1603] [General] [answered]
## Extension exclusion alone works
`rg -i "foo" --glob '{*.js}' --glob '!{*.min.js}' -l`
stuff.js
stuff2.js
foo/my-lib/a.js
foo/my-lib/b.js

## Not anymore when also excluding directories
`rg -i "foo" --glob '{*.js}' --glob '!{*.min.js,foo/my-lib/*}' -l`
`rg -i "foo" --glob '{*.js}' --glob '!{foo/my-lib/*,*.min.js}' -l`
stuff.js
stuff2.js
my-other-lib.min.js

**Comment by tuxayo** [comment]
Maybe it's a bug but I'm not sure if that it should work. It's likely that I misunderstood the doc.

**Comment by tuxayo** [comment]
I'm not sure if --type-add could be of any help.

https://www.mankier.com/1/rg#--type-add

**Comment by BurntSushi** [comment, marked as answer]
Hmm, I cannot reproduce:

```
$ mkdir -p foo/my-lib

$ touch stuff.js stuff2.js my-other-lib.min.js foo/my-lib/{a,b}.js

$ rg --files
stuff.js
stuff2.js
my-other-lib.min.js
foo/my-lib/b.js
foo/my-lib/a.js

$ rg --files -g '*.js' -g '!*.min.js'
stuff.js
stuff2.js
foo/my-lib/b.js
foo/my-lib/a.js

$ rg --files -g '*.js' -g '!{*.min.js,foo/my-lib/*}'
stuff.js
stuff2.js

$ rg --files -g '{*.js}' -g '!{*.min.js,foo/my-lib/*}'
stuff.js
stuff2.js

$ rg --files -g '{*.js}' -g '!{foo/my-lib/*,*.min.js}'
stuff.js
stuff2.js
```

A few questions for you:

1. What version of ripgrep are you using? How did you install it?
2. Can you try running the same commands as above, but with the `--debug` flag?

**Reply by tuxayo** [reply]
##### 1. I can't reproduce with your example
it works as expected. :o

##### 2. I can't reproduce if I move some files from my project to a dedicated directory
to have a "mini-case" based on my full project. It works as expected :o

##### 3. Let's try with the full project
Only 40 MiB of download, git does a great job!
```
git clone --depth 1 https://gitlab.com/koha-community/Koha.git
cd Koha
```
I can reproduce!

```
/t/m/Koha ❯❯❯ rg --files --glob '{*.js}' --glob '!{koha-tmpl/intranet-tmpl/lib/jquery/*,*.min.js}' | wc -l
272
# *.min.js files still in the list

/t/m/Koha ❯❯❯ rg --files --glob '{*.js}' --glob '!{*.min.js}' | wc -l
216
# koha-tmpl/intranet-tmpl/lib/jquery/* files still in the list

/t/m/Koha ❯❯❯ rg --files --glob '{*.js}' --glob '!{koha-tmpl/intranet-tmpl/lib/jquery/*}' | wc -l
272
# *.min.js files still in the list
```


```
/t/m/Koha ❯❯❯ rg --version
ripgrep 12.1.0
-SIMD -AVX (compiled)
+SIMD -AVX (runtime)

```

Installed via the Arch Linux package.

**Reply by BurntSushi** [reply]
Ah great, thank you. That led me to a smaller reproduction based on your original example:

```
$ tree
.
├── bar
│   └── my-other-lib.min.js
├── foo
│   └── my-lib
│       ├── a.js
│       └── b.js
├── my-other-lib.min.js
├── stuff2.js
└── stuff.js

3 directories, 6 files

$ rg --files -g '{*.js}' -g '!{foo/my-lib/*,*.min.js}'
stuff.js
stuff2.js
bar/my-other-lib.min.js

$ rg --files -g '{*.js}' -g '!{foo/my-lib/*}' -g '!*.min.js'
stuff2.js
stuff.js
```

While this might seem strange, this is actually correct behavior. The key issue here is that the `-g/--glob` flag is essentially a command line interface to specifying `gitignore` globs, and those globs adhere fairly strictly to the specification included in `man gitignore`. In particular, if a glob pattern contains a `/` _anywhere_, then the glob must match the full pattern. If there is no `/`, then a `*` can itself match through a `/`. Otherwise it can't. For example, `foo/*.min.js` will match `foo/test.min.js` but it _won't_ match `foo/bar/test.min.js`. Instead, you'd have to use a recursive glob, e.g., `foo/**/*.min.js`.

So in your particular example, your glob is `!{foo/my-lib/*,*.min.js}`. The `/` in that glob means that a `*` cannot match through a `/`. So this is why it excludes `my-other-lib.min.js`, which has no `/` in it. But you'll notice that it _doesn't_ exclude `bar/my-other-lib.js`, because the `*` in `*.min.js` cannot match through the `/` after `bar`.

So the right way to work around this is to just specify another glob. Instead of `-g '!{foo/my-lib/*,*.min.js}`, you want `-g '!foo/my-lib/*' -g '!*.min.js'`.

N.B. Strictly speaking, this is actually a bit weirder than gitignore, since gitignore doesn't actually support the curly brace syntax. One could then make an argument that a `/` in one of the patterns in a brace expansion shouldn't impact the semantics of another pattern in the same brace. That indeed might be better.

**Reply by tuxayo** [reply]
> those globs adhere fairly strictly to the specification included in man gitignore

Thanks, I wasn't sure of that, that's why I didn't dig deeper than ripgrep's doc.

> So the right way to work around this is to just specify another glob. Instead of -g '!{foo/my-lib/*,*.min.js}, you want -g '!foo/my-lib/*' -g '!*.min.js'.

Thanks for your explanation and your help. I'll be able to keep enjoying using `rg` for my current task :smile:
